#include <iostream>
#include <memory>
#include <string>

#include "shop.grpc.pb.h"
#include <grpcpp/grpcpp.h>

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;

using shop::CheckStatusRequest;
using shop::CheckStatusResponse;
using shop::OrderStatus;
using shop::PlaceOrderRequest;
using shop::PlaceOrderResponse;
using shop::ShopService;

class ShopClient {
public:
  ShopClient(std::shared_ptr<Channel> channel)
      : stub_(ShopService::NewStub(channel)) {}

  int PlaceOrder(const std::string &size, const std::string &color,
                 const std::string &text) {
    PlaceOrderRequest request;
    request.set_size(size);
    request.set_color(color);
    request.set_text(text);

    PlaceOrderResponse response;
    ClientContext context;

    Status status = stub_->PlaceOrder(&context, request, &response);
    if (status.ok()) {
      std::cout << "‚úÖ –ó–∞–∫–∞–∑ –æ—Ñ–æ—Ä–º–ª–µ–Ω. –ù–æ–º–µ—Ä –∑–∞–∫–∞–∑–∞: " << response.order_id()
                << std::endl;
      return response.order_id();
    } else {
      std::cerr << "‚ùå –û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–∫–∞–∑–µ: " << status.error_message()
                << std::endl;
      return -1;
    }
  }

  void CheckStatus(int order_id) {
    CheckStatusRequest request;
    request.set_order_id(order_id);

    CheckStatusResponse response;
    ClientContext context;

    Status status = stub_->CheckStatus(&context, request, &response);

    if (status.ok()) {
      std::cout << "üì¶ –°—Ç–∞—Ç—É—Å –∑–∞–∫–∞–∑–∞ #" << order_id << ": "
                << StatusToString(response.status()) << std::endl;
    } else {
      std::cerr << "‚ùå –û—à–∏–±–∫–∞ –ø—Ä–∏ –ø–æ–ª—É—á–µ–Ω–∏–∏ —Å—Ç–∞—Ç—É—Å–∞: " << status.error_message()
                << std::endl;
    }
  }

private:
  std::unique_ptr<ShopService::Stub> stub_;

  std::string StatusToString(OrderStatus status) {
    switch (status) {
    case OrderStatus::MAKING:
      return "–î–µ–ª–∞–µ—Ç—Å—è";
    case OrderStatus::IN_TRANSIT:
      return "–í –ø—É—Ç–∏";
    case OrderStatus::DELIVERED:
      return "–î–æ—Å—Ç–∞–≤–ª–µ–Ω";
    default:
      return "–ù–µ–∏–∑–≤–µ—Å—Ç–Ω–æ";
    }
  }
};

void PrintMenu() {
  std::cout << "\n--- –ú–µ–Ω—é ---\n"
            << "1. –ó–∞–∫–∞–∑–∞—Ç—å —Ñ—É—Ç–±–æ–ª–∫—É\n"
            << "2. –ü—Ä–æ–≤–µ—Ä–∏—Ç—å —Å—Ç–∞—Ç—É—Å –∑–∞–∫–∞–∑–∞\n"
            << "0. –í—ã—Ö–æ–¥\n"
            << "–í—ã–±–µ—Ä–∏—Ç–µ –¥–µ–π—Å—Ç–≤–∏–µ: ";
}

int main() {
  auto address = std::string("localhost:50051");
  ShopClient client(
      grpc::CreateChannel(address, grpc::InsecureChannelCredentials()));

  int choice;
  while (true) {
    PrintMenu();
    std::cin >> choice;
    std::cin.ignore(); // –æ—á–∏—Å—Ç–∫–∞ –±—É—Ñ–µ—Ä–∞

    if (choice == 0) {
      std::cout << "–í—ã—Ö–æ–¥...\n";
      break;
    }

    if (choice == 1) {
      std::string size, color, text;

      std::cout << "–†–∞–∑–º–µ—Ä —Ñ—É—Ç–±–æ–ª–∫–∏ (–Ω–∞–ø—Ä–∏–º–µ—Ä, M, L, XL): ";
      std::getline(std::cin, size);

      std::cout << "–¶–≤–µ—Ç —Ñ—É—Ç–±–æ–ª–∫–∏: ";
      std::getline(std::cin, color);

      std::cout << "–ù–∞–¥–ø–∏—Å—å: ";
      std::getline(std::cin, text);

      client.PlaceOrder(size, color, text);
    } else if (choice == 2) {
      int order_id;
      std::cout << "–í–≤–µ–¥–∏—Ç–µ –Ω–æ–º–µ—Ä –∑–∞–∫–∞–∑–∞: ";
      std::cin >> order_id;
      std::cin.ignore();

      client.CheckStatus(order_id);
    } else {
      std::cout << "‚ùó –ù–µ–≤–µ—Ä–Ω—ã–π –≤—ã–±–æ—Ä\n";
    }
  }

  return 0;
}
